pkgs/desktops/gnome-3/extensions/update-extensions.py

```diff
@@ -0,0 +1,247 @@
+                    str: { "version": int, "sha256": str },
+                    …
+                }
+            }
+    """
+    uuid = extension["uuid"]
+
+    # Yeah, there are some extensions without any releases
+    if not extension["shell_version_map"]:
+        return None
+    logging.info("Processing '" + uuid + "'")
+
+    # Input is a mapping str -> { version: int, … }
+    # We want to map shell versions to extension versions
+    shell_version_map: Dict[str, int] = {
+        k: v["version"] for k, v in extension["shell_version_map"].items()
+    }
+    # Transform shell_version_map to be more useful for us. Also throw away unwanted versions
+    shell_version_map: Dict[str, Dict[str, str]] = generate_extension_versions(shell_version_map, uuid)  # type: ignore
+
+    # No compatible versions found
+    if not shell_version_map:
+        return None
+
+    # Fetch a human-readable name for the package. If it is already taken by some package,
+    # append an ID number to make it unique. This is the first thing we have to do, as the filtering
+    # may change over time and thus result in extensions being retroactively renamed. Which we don't want.
+    (pname, pname_id) = pname_from_url(extension["link"])
+
+    for shell_version in shell_version_map.keys():
+        if pname in package_name_registry[shell_version]:
+            logging.warning(f"Package name '{pname}' is colliding.")
+            package_name_registry[shell_version][pname] += [uuid]
+        else:
+            package_name_registry[shell_version][pname] = [uuid]
+
+    return {
+        "uuid": uuid,
+        "name": sanitize_string(extension["name"]),
+        "pname": pname,
+        "description": sanitize_string(extension["description"]),
+        "link": "https://extensions.gnome.org" + extension["link"],
+        "shell_version_map": shell_version_map,
+    }
+
+
+def fetch_extensions() -> List[Dict[str, Any]]:
+    """
+    Fetch the extensions. The extensions database is scraped by using the search query. We simply go over it
+    page by page until we hit a non-full page or a 404 error.
+
+    The returned list is sorted by the age of the extension, in order to be deterministic.
+    """
+    page = 0
+    extensions = []
+    while True:
+        page += 1
+        logging.info("Scraping page " + str(page))
+        try:
+            with urllib.request.urlopen(
+                f"https://extensions.gnome.org/extension-query/?n_per_page=25&page={page}"
+            ) as url:
+                data = json.loads(url.read().decode())["extensions"]
+                responseLength = len(data)
+
+                for extension in data:
+                    extensions += [extension]
+
+                # If our page isn't "full", it must have been the last one
+                if responseLength < 25:
+                    logging.debug(
+                        f"\tThis page only has {responseLength} entries, so it must be the last one."
```

Comment:
Logging should never use f-strings. The idea is that the expensive operation of substituting some strings in there should only be done if the log level requires it and otherwise it should be constant cost to skip this line.

I'll not be pedantic on this here as well this isn't a real application anyway...

Reply:
Thanks for the notice, but I think changing this would not improve the readability of the code, and the performance impact of this is absolutely negligible.

Reply:
It isn't about performance but consistent pythonic code style.